dsgrid/query/derived_dataset.py

```diff
@@ -0,0 +1,265 @@
+                dimension_mapping_references.append(
+                    _get_dimension_mapping_reference(supp_dim, project.config)
+                )
+
+    if dimension_mapping_references:
+        _make_dimension_mapping_references_file(dimension_mapping_references, dst_path)
+
+    _make_dataset_config(
+        query.project.dataset.dataset_id,
+        metadata.pivoted.dimension_type.value,
+        dimensions,
+        dimension_references,
+        dst_path,
+    )
+    return True
+
+
+def does_query_support_a_derived_dataset(query: ProjectQueryModel):
+    """Return True if a derived dataset can be created from a query.
+
+    Returns
+    -------
+    bool
+    """
+    is_valid = True
+    if query.result.column_type != ColumnType.DIMENSION_TYPES:
+        is_valid = False
+        logger.error(
+            "Cannot create a derived dataset from a query with column_type = %s. It must be %s.",
+            query.result.column_type.value,
+            ColumnType.DIMENSION_TYPES.value,
+        )
+    if query.result.supplemental_columns:
+        is_valid = False
+        logger.error("Cannot create a derived dataset from a query with supplemental_columns")
+    if query.result.replace_ids_with_names:
+        is_valid = False
+        logger.error("Cannot create a derived dataset from a query with replace_ids_with_names")
+
+    return is_valid
+
+
+def _is_dimension_valid_for_dataset(dim_config, unique_data_records):
+    records = dim_config.get_records_dataframe()
+    dim_values = get_unique_values(records, "id")
+    diff = dim_values.symmetric_difference(unique_data_records)
+    return not bool(diff)
+
+
+def _get_matching_supplemental_dimension(project_config, dimension_type, unique_data_records):
+    for dim_config in project_config.list_supplemental_dimensions(dimension_type):
+        if _is_dimension_valid_for_dataset(dim_config, unique_data_records):
+            return dim_config.model
+
+    return None
+
+
+def _make_dataset_config(
+    dataset_id,
+    pivoted_dim_type,
+    dimensions,
+    dimension_references,
+    path: Path,
+    data_classification=DataClassificationType.MODERATE.value,
+):
+    config = {
+        "dataset_id": dataset_id,
+        "dataset_type": InputDatasetType.MODELED.value,
+        "data_source": "",
+        "data_schema_type": DataSchemaType.ONE_TABLE.value,
+        "data_schema": {
+            "load_data_column_dimension": pivoted_dim_type,
+        },
+        "dataset_version": "1.0.0",
+        "description": "",
+        "origin_creator": "",
+        "origin_organization": "",
+        "origin_date": "",
+        "origin_project": "",
+        "origin_version": "",
+        "source": "",
+        "data_classification": data_classification,
+        "use_project_geography_time_zone": True,
+        "dimensions": dimensions,
+        "dimension_references": dimension_references,
+    }
```

Comment:
Seems like it would be safer to create using the pydantic data model classes?

Reply:
I did this with dictionaries because it skips validation that would otherwise fail in the dimension section (because of record checks that are invalid in this context). I copied the comment from below here. I don't think it's any less safe. If any required field is missing, registration will fail.

Note that since we now may skip adding a time dimension, more checks will fail if we use the Pydantic models.